validators/AgeValidator.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace humhub\modules\legal\validators;
+
+use DateTime;
+use Yii;
+use yii\validators\Validator;
+use humhub\modules\user\models\User;
+use humhub\modules\user\models\Group;
+
+/**
+ * AgeValidator validates that the given value represents an age greater than or equal to a specified minimum age.
+ */
+class AgeValidator extends Validator
+{
+    /**
+     * @var int The minimum age required
+     */
+    public $minimumAge;
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if ($this->minimumAge === null) {
+            $this->minimumAge = Yii::$app->getModule('legal')->getMinimumAge();
+        }
+    }
+
+    /**
+     * Validates the age of the user based on the given attribute value.
+     *
+     * @param \yii\base\Model $model the data model being validated
+     * @param string $attribute the name of the attribute to be validated
+     */
+    public function validateAttribute($model, $attribute)
+    {
+        // Check if the user is a member of the admin group
+        if (Group::getAdminGroup()->isMember($model)) {
+            // Skip validation for admin accounts
+            return;
+        }
+
+        $value = $model->$attribute;
+
+        if (!$value instanceof DateTime) {
+            try {
+                $value = new DateTime($value);
+            } catch (\Exception $e) {
+                $this->addError($model, $attribute, Yii::t('LegalModule.base', 'Invalid date format.'));
+                return;
+            }
+        }
+
+        $today = new DateTime();
+        $age = $today->diff($value)->y;
+
+        if ($age < $this->minimumAge) {
+            $message = Yii::t('LegalModule.base', 'You must be at least {age} years old.', ['age' => $this->minimumAge]);
+            $this->addError($model, $attribute, $message);
+
+            if ($this->minimumAge > 0 && isset($model->user) && $model->user instanceof User) {
+                $model->user->status = User::STATUS_DISABLED;
+                $model->user->save(false);
```

Comment:
@ArchBlood Do you know why we have the redirect after disabling user?
I mean the following strange behaviour:

- I save user with age < 16 years
- Click save
- I see the form page is loaded with red error `You must be at least 16 years old.` under the birthday input, but I see the form only 1-2 seconds and then I have a refirect to a login page, so I even cannot fix the error.

You can try to add here a code like `\Yii::$app->view->error($message);` but the status bottom error is displayed only on the user edit form for 1-2 seconds, but after redirect I don't see it on the login page, so it is so few time to read it.

If need I can record a video to display how it works, but I think user should have a possibility to fix the birth date if he enters it by mistake before the unexpecting disabling and logout.

Reply:
@ArchBlood 

https://github.com/user-attachments/assets/a832ea5e-a32b-469e-893b-e3867839bb7f

I can get this with admin and normal user on browser Chrome and FireFox on Windows on v1.16 and v1.17
If you cannot reproduce this on your side then I guess the action `/live/poll` is called and probably some external module does the auto-redirect of a disabled user.
I will review it deeply later if you cannot find the reason of the redirect.

Reply:
@yurabakhtin I've tested on switching an account that is not admin and one that is admin, on save it does disable the account that isn't currently saved as an admin but no redirect happens just the highlighted date field for "Birthday" with the error message till fixed so on this issue I do not see anything with the current version of the PR;

### Normal User Test
![Screenshot_1](https://github.com/user-attachments/assets/9aa9b5d8-7a78-4557-baa9-0daf5eec16fb)

### Admin User Test
![Screenshot_2](https://github.com/user-attachments/assets/e61c3187-850b-4686-a5ee-6f585a410de0)

Reply:
If there is an issue then my only guess would have to be within [UserProfileController.php](https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/controllers/UserProfileController.php) otherwise I don't have the issue.

Reply:
In https://github.com/humhub/legal/pull/89/commits/b00d78355153b0f5b31616753a19b0fc6c12fc69 I've updated the validation so that it will show the error message for either admin or normal user on the admin settings when editing a user's account, it should also still work for when a guest registers a new account like it did before and admin accounts should not be affected, but I'm still unable to reproduce the issue reported in https://github.com/humhub/legal/pull/89#discussion_r1773640184.

Reply:
@ArchBlood Yes, it works like you show but it is only for Administration side, but I tell about case when you are logged in as the user and update it from Account menu -> Settings -> Profile -> General, please try it again.

Reply:
> @ArchBlood Yes, it works like you show but it is only for Administration side, but I tell about case when you are logged in as the user and update it from Account menu -> Settings -> Profile -> General, please try it again.

After reviewing Account menu -> Settings -> Profile -> General, I do see what seems to be a bug, but I'm not sure if it's something to do with this P/R to be exact, as now I do see a redirect action happening, but the redirect is not to the login form but the `manifest.json`  which I'm not sure how this is currently possible, my only guess is that it happens on impersonation of another user? A possible theory could be the use of `onBeforeValidate()`? Other than this I'm rather stumped on why this is happening. 🤔

![Screenshot_3](https://github.com/user-attachments/assets/a789bb87-d582-477c-8395-61a212aa77ae)

Reply:
@ArchBlood I cannot find what code does the redirect.
But why should we disable an user right in the validator code, it seems it is not a correct place, I think the validator sohuld be used only in order to add errors to the model so it is restricted to be saved.
Maybe the code should be moved to another event like `afterSave`?

What if instead of the user disabling we will redirect a logged in user to some page where we will inform about the age issue?
I mean a page like the module displays on the route `/legal/page/confirm` if the user still didn't accept terms.

Reply:
> @ArchBlood I cannot find what code does the redirect.
> But why should we disable an user right in the validator code, it seems it is not a correct place, I think the validator sohuld be used only in order to add errors to the model so it is restricted to be saved.
> Maybe the code should be moved to another event like `afterSave`?
> 
> What if instead of the user disabling we will redirect a logged in user to some page where we will inform about the age issue?
> I mean a page like the module displays on the route `/legal/page/confirm` if the user still didn't accept terms.

I'm not sure we should introduce more redirects here, in the P/R before this one the user was simply force logged out due to making their birthday not within the age limit, my thinking was, who doesn't know what their birthday is when it's requested, if they didn't meet age requirements then they shouldn't need a second chance because they probably shouldn't be there, it was then discussed that maybe disabling would be better, which kind of works the same way, we could simplify it and just remove the disable part all together but keep the error message as is?

Reply:
@yurabakhtin I've removed disabling a user's account in https://github.com/humhub/legal/pull/89/commits/6ac159935a010f262ef49f94613d90f3a29e0654, as for the error message, I've tried using `\Yii::$app->view->error($message);` which doesn't seem to work whereas `$this->addError($model, $attribute, $message);` does. See Yii's [docs](https://www.yiiframework.com/doc/api/2.0/yii-validators-validator#addError()-detail) for the reason behind this.

I now no longer see any redirects, so my guess is that this is an issue when `$model->user->status = User::STATUS_DISABLED` is used. I remember running into a similar issue before but for the life of me I can't remember which repo the issue was reported in. 🤔

Reply:
> I'm not sure we should introduce more redirects here, in the P/R before this one the user was simply force logged out due to making their birthday not within the age limit, my thinking was, who doesn't know what their birthday is when it's requested, if they didn't meet age requirements then they shouldn't need a second chance because they probably shouldn't be there, it was then discussed that maybe disabling would be better, which kind of works the same way, we could simplify it and just remove the disable part all together but keep the error message as is?

@ArchBlood Ah sorry if this behaviour was accepted by some discussion, then you should keep the deleted `$model->user->status = User::STATUS_DISABLED;`.
I just thought we should give a second chance to fix the birthday error.

@luke- I can approve this PR if you agree the behaviour so when user enters a wrong birthday and then it is logged out automatically.

Reply:
@ArchBlood @yurabakhtin 

I don't like the solution with an automatic logout. I would just display a validation message from the validator that the value is not valid.

Reply:
> @ArchBlood @yurabakhtin
> 
> I don't like the solution with an automatic logout. I would just display a validation message from the validator that the value is not valid.

I have removed disabling the account here https://github.com/humhub/legal/pull/89/commits/6ac159935a010f262ef49f94613d90f3a29e0654, in the future we should think of making a soft disable in these kind of situations which is more ideal for this kind of feature, but overall the P/R should be complete, all that needs reviewed is the tests which I'm not sure are correct in this case as I normally don't work on these myself.

Reply:
@ArchBlood You removed only the line `$model->user->status = User::STATUS_DISABLED;` but then you should remove the rest code:

```php
            // Disable the user account if they are underage and not an admin
            if ($this->minimumAge > 0 && isset($model->user) && $model->user instanceof User && !$model->user->isSystemAdmin()) {
                $model->user->save(false);
            }
```

because I don't understand why we need to update the user if none of the columns is updated.

Reply:
@yurabakhtin Done in https://github.com/humhub/legal/pull/89/commits/1a4c9fbb7fb5dbf93e448e2a15c61422cffb2b32